File: nodejs-typescript/src/application/telemetry/TransferProcessedTelemetry.ts
import type { TransferResult } from '../../domain/TransferResult.js';
import {
  isTransferSkipped,
  isTransferSuccess,
  transferReasonCode,
} from '../../domain/TransferResult.js';

export type TransferProcessedLogEntry = {
  readonly event: typeof TransferProcessedTelemetry.EVENT_NAME;
  readonly outcome: string;
  readonly reason_code: string | null;
  readonly transfer_id: string;
  readonly from: string;
  readonly to: string;
};

export type MetricIncrement = {
  readonly name: typeof TransferProcessedTelemetry.METRIC_NAME;
  readonly outcome: string;
  readonly reason: string | null;
};

export class TransferProcessedTelemetry {
  static readonly METRIC_NAME = 'transfer.processed' as const;
  static readonly EVENT_NAME = 'transfer.processed' as const;

  private constructor(
    readonly outcome: string,
    readonly reasonCode: string | null,
    readonly transferId: string,
    readonly from: string,
    readonly to: string,
  ) {}

  static fromResult(result: TransferResult): TransferProcessedTelemetry {
    const outcome = isTransferSuccess(result)
      ? 'succeeded'
      : isTransferSkipped(result)
        ? 'skipped'
        : 'failed';
    return new TransferProcessedTelemetry(
      outcome,
      transferReasonCode(result),
      result.instruction.transferId.toString(),
      result.instruction.fromAccountNumber.toString(),
      result.instruction.toAccountNumber.toString(),
    );
  }

  toLogEntry(): TransferProcessedLogEntry {
    return {
      event: TransferProcessedTelemetry.EVENT_NAME,
      outcome: this.outcome,
      reason_code: this.reasonCode,
      transfer_id: this.transferId,
      from: this.from,
      to: this.to,
    };
  }

  toMetricIncrement(): MetricIncrement {
    return {
      name: TransferProcessedTelemetry.METRIC_NAME,
      outcome: this.outcome,
      reason: this.reasonCode,
    };
  }
}

File: ruby/lib/mable_bank/application/telemetry/transfer_processed_telemetry.rb
# frozen_string_literal: true

module MableBank
  module Application
    module Telemetry
      TransferProcessedLogEntry = Data.define(:event, :outcome, :reason_code, :transfer_id, :from, :to)
      MetricIncrement = Data.define(:name, :outcome, :reason)

      class TransferProcessedTelemetry
        METRIC_NAME = 'transfer.processed'
        EVENT_NAME = 'transfer.processed'

        attr_reader :outcome, :reason_code, :transfer_id, :from, :to

        def self.from_result(result)
          outcome = if result.success?
                      'succeeded'
                    elsif result.skipped?
                      'skipped'
                    else
                      'failed'
                    end
          new(
            outcome: outcome,
            reason_code: result.reason_code&.to_s,
            transfer_id: result.instruction.transfer_id.to_s,
            from: result.instruction.from_account_number.to_s,
            to: result.instruction.to_account_number.to_s
          )
        end

        def initialize(outcome:, reason_code:, transfer_id:, from:, to:)
          @outcome = outcome
          @reason_code = reason_code
          @transfer_id = transfer_id
          @from = from
          @to = to
        end

        def to_log_entry
          TransferProcessedLogEntry.new(
            event: EVENT_NAME,
            outcome: @outcome,
            reason_code: @reason_code,
            transfer_id: @transfer_id,
            from: @from,
            to: @to
          )
        end

        def to_metric_increment
          MetricIncrement.new(name: METRIC_NAME, outcome: @outcome, reason: @reason_code)
        end
      end
    end
  end
end
